modules/mod-deep-learning/DeepLearningAnalyzer.cpp

```diff
@@ -0,0 +1,232 @@
+   // the length of the selection being processed.
+   Floats buffer{leader->GetMaxBlockSize()};
+
+   // get each of the blocks we will process
+   for (BlockIndex block : GetBlockIndices(leader, tStart, tEnd)) 
+   {
+      // Get a blockSize of samples (smaller than the size of the buffer)
+      sampleCount samplePos = block.first;
+      size_t blockSize = block.second;
+
+      // get a torch tensor from the leader track
+      torch::Tensor input =
+         BuildMultichannelTensor(leader, buffer.get(), samplePos, blockSize);
+
+      // resample!
+      input = mModel->Resample(input, origRate, mModel->GetSampleRate());
+
+      // if we're not doing a multichannel forward pass, downmix
+      if (!mModel->GetCard()->multichannel())
+         input = input.sum(0, true, torch::kFloat);
+
+      // forward pass!
+      torch::jit::IValue output = ForwardPassInThread(input);
+
+      // split forward pass output into output tensor and timestamps
+      auto [modelOutput, timestamps] = mModel->ToTimestamps(output);
+
+      // write the block's label to the label track
+      double blockStart = leader->LongSamplesToTime(samplePos);
+      sampleCount blockEndSamples = samplePos + (sampleCount)blockSize;
+      double blockEnd = leader->LongSamplesToTime(blockEndSamples);
+
+      TensorToLabelTrack(modelOutput, labelTrack, blockStart, blockEnd, timestamps);
+
+      // Update the Progress meter
+      double tPos = leader->LongSamplesToTime(samplePos);
+      if (TrackProgress(mCurrentTrackNum, (tPos - tStart) / (tEnd - tStart)))
+      return false;
+   }
+   labelTrack->Commit();
+   return true;
+}
+
+std::vector<Stamp> Coalesce(std::vector<Stamp> &timestamps)
+{
+   if (timestamps.size() == 0)
+      return timestamps;
+
+   // sort the stamps by start time
+   std::function< bool(Stamp &a, Stamp &b)> compareByStartTime (
+      [](Stamp &a, Stamp &b)
+      {
+         return a.first < b.first;
+      }
+   );
+
+   std::sort(timestamps.begin(), timestamps.end(), compareByStartTime);
+   std::vector<Stamp> coalescedStamps; 
+
+   Stamp prev = timestamps[0];
+   for (const auto &stamp : timestamps)
+   {
+      if (stamp.first <= prev.second)
+      {
+         if (stamp.second > prev.second)
+         {
+            prev.second = stamp.second;
+         }
+      }
+      else
+      {
+         coalescedStamps.emplace_back(Stamp(prev));
+         prev = Stamp(stamp);
+      }
+   }
+
+   // check if last is not equal to end of last stamp
+   // append the last stamp
+   if (!coalescedStamps.empty())
+   {
+      if (coalescedStamps.back().second != prev.second)
+      {
+         coalescedStamps.emplace_back(prev);
+      }
+   }
+   else if (coalescedStamps.empty() && !timestamps.empty())
+   {
+      coalescedStamps.emplace_back(prev);
+   }
+
+   return coalescedStamps;
+}
+
+// TODO: coalesce labels
+void DeepLearningAnalyzer::TensorToLabelTrack
+(torch::Tensor output, std::shared_ptr<AddedAnalysisTrack> labelTrack,
```

Comment:
I would pass labelTrack by reference, since the smart pointer ownership policy is not relevant to what the function does.

Reply:
3db261fb8644bbb764504db69d4fd10223f7dec0